# Lab-Based Quiz 4  
# Time: 1 hour
## Comparing Individual Models vs Ensemble Methods  
### Dataset: Breast Cancer Wisconsin (Diagnostic)


## Objective

In this lab, you will:

1. Train and evaluate the following classifiers individually:
   - Decision Tree
   - Naive Bayes
   - K-Nearest Neighbors (KNN)
   - Logistic Regression
   - SGD Classifier

2. Build ensemble models using:
   - Hard Voting
   - Soft Voting
   - Bagging
   - Random Forest
   - AdaBoost

3. Compare all models using appropriate evaluation metrics.

4. Decide which metric best represents performance in this case study and justify your reasoning.

5. Write your own summary explaining your findings.

**Important**:
- Do NOT hard-code results.
- Do NOT use Gemini or any LLM model to do this for you.
- All answers must be written in your own words.
- Replace each TODO with working code.

# Part 1 – Load and Explore the Dataset

The Breast Cancer Wisconsin dataset is a binary classification dataset.

In [1]:
# TODO: Import necessary libraries

# TODO: Load the breast cancer dataset from sklearn.datasets

# TODO: Convert the dataset to a pandas DataFrame (features + target)

# TODO: Display the first 5 rows of the dataset

# TODO: Import necessary libraries
import pandas as pd
from sklearn.datasets import load_breast_cancer

# TODO: Load the breast cancer dataset from sklearn.datasets
cancer_data = load_breast_cancer()

# TODO: Convert the dataset to a pandas DataFrame (features + target)
df = pd.DataFrame(cancer_data.data, columns=cancer_data.feature_names)
df['target'] = cancer_data.target

# TODO: Display the first 5 rows of the dataset
df.head()

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


## Basic Exploration

1. How many samples and features are there?
2. What are the target classes?
3. Is the dataset balanced?

In [3]:
# TODO: Print dataset shape
print("Dataset shape (Rows, Columns):", df.shape)

# TODO: Print class distribution
print("\nClass Distribution:")
print(df['target'].value_counts())

# TODO: Compute percentage of each class
print("\nPercentage of each class:")
print(df['target'].value_counts(normalize=True) * 100)

Dataset shape (Rows, Columns): (569, 31)

Class Distribution:
target
1    357
0    212
Name: count, dtype: int64

Percentage of each class:
target
1    62.741652
0    37.258348
Name: proportion, dtype: float64


### Write your answers below:

- Number of samples: 569
- Number of features: 30 (plus 1 target column)
- Class distribution: 357 Benign (1), 212 Malignant (0)
- Is the dataset balanced?
  - No, it is imbalanced. There is a clear majority class (Benign at ~63%) compared to the minority class (Malignant at ~37%).

# Part 2 – Train/Test Split and Scaling

In [4]:
# TODO: Perform train_test_split
# - 80% training
# - 20% testing
# - random_state=42

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X = df.drop('target', axis=1)
y = df['target']
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [6]:
# TODO: Apply StandardScaler
scaler = StandardScaler()

# TODO: Fit scaler on training data only
scaler.fit(X_train)

# TODO: Transform both training and test sets
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

### In 2–3 sentences:
Why is scaling necessary for some models but not for others?

- Scaling is essential for distance-based models like KNN, which perform best when all features are on the same scale, as features with larger numerical ranges will disproportionately dominate distance calculations and bias the results. It is also necessary for gradient-based models (like Logistic Regression) to ensure the optimization process converges efficiently. In contrast, tree-based models are unaffected by scale because they partition data using relative thresholds rather than calculating magnitudes or distances.

# Part 3 – Train Individual Models

For EACH model:

* Train

* Predict

* Compute:

  * Accuracy

  * Precision

  * Recall

  * F1-score

  * Confusion Matrix

Store results for later comparison

In [7]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# TODO: Import DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier

# TODO: Initialize model
dt_model = DecisionTreeClassifier(random_state=42)

# TODO: Fit model
dt_model.fit(X_train_scaled, y_train)

# TODO: Predict
dt_preds = dt_model.predict(X_test_scaled)

# TODO: Compute evaluation metrics
dt_results = {
    "Accuracy": accuracy_score(y_test, dt_preds),
    "Precision": precision_score(y_test, dt_preds),
    "Recall": recall_score(y_test, dt_preds),
    "F1-score": f1_score(y_test, dt_preds),
    "Confusion Matrix": confusion_matrix(y_test, dt_preds)
}

# TODO: Store results in a dictionary
results_dict = {"Decision Tree": dt_results}

In [8]:
# TODO: Import GaussianNB
from sklearn.naive_bayes import GaussianNB

# TODO: Initialize model
nb_model = GaussianNB()

# TODO: Fit model
nb_model.fit(X_train_scaled, y_train)

# TODO: Predict
nb_preds = nb_model.predict(X_test_scaled)

# TODO: Compute evaluation metrics
nb_results = {
    "Accuracy": accuracy_score(y_test, nb_preds),
    "Precision": precision_score(y_test, nb_preds),
    "Recall": recall_score(y_test, nb_preds),
    "F1-score": f1_score(y_test, nb_preds),
    "Confusion Matrix": confusion_matrix(y_test, nb_preds)
}

# TODO: Store results
results_dict["Naive Bayes"] = nb_results

In [9]:
# TODO: Import KNeighborsClassifier
from sklearn.neighbors import KNeighborsClassifier

# TODO: Initialize model
knn_model = KNeighborsClassifier()

# TODO: Fit model
knn_model.fit(X_train_scaled, y_train)

# TODO: Predict
knn_preds = knn_model.predict(X_test_scaled)

# TODO: Compute evaluation metrics
knn_results = {
    "Accuracy": accuracy_score(y_test, knn_preds),
    "Precision": precision_score(y_test, knn_preds),
    "Recall": recall_score(y_test, knn_preds),
    "F1-score": f1_score(y_test, knn_preds),
    "Confusion Matrix": confusion_matrix(y_test, knn_preds)
}

# TODO: Store results
results_dict["KNN"] = knn_results

In [10]:
# TODO: Import LogisticRegression
from sklearn.linear_model import LogisticRegression

# TODO: Initialize model
lr_model = LogisticRegression(random_state=42)

# TODO: Fit model
lr_model.fit(X_train_scaled, y_train)

# TODO: Predict
lr_preds = lr_model.predict(X_test_scaled)

# TODO: Compute evaluation metrics
lr_results = {
    "Accuracy": accuracy_score(y_test, lr_preds),
    "Precision": precision_score(y_test, lr_preds),
    "Recall": recall_score(y_test, lr_preds),
    "F1-score": f1_score(y_test, lr_preds),
    "Confusion Matrix": confusion_matrix(y_test, lr_preds)
}

# TODO: Store results
results_dict["Logistic Regression"] = lr_results

In [11]:
# TODO: Import SGDClassifier
from sklearn.linear_model import SGDClassifier

# TODO: Initialize model
sgd_model = SGDClassifier(random_state=42)

# TODO: Fit model
sgd_model.fit(X_train_scaled, y_train)

# TODO: Predict
sgd_preds = sgd_model.predict(X_test_scaled)

# TODO: Compute evaluation metrics
sgd_results = {
    "Accuracy": accuracy_score(y_test, sgd_preds),
    "Precision": precision_score(y_test, sgd_preds),
    "Recall": recall_score(y_test, sgd_preds),
    "F1-score": f1_score(y_test, sgd_preds),
    "Confusion Matrix": confusion_matrix(y_test, sgd_preds)
}

# TODO: Store results
results_dict["SGD Classifier"] = sgd_results

In [41]:
import pandas as pd

model_results_list = []

for model_name, metrics in results_dict.items():
    model_results_list.append({
        "Model": model_name,
        "Accuracy": metrics["Accuracy"],
        "Precision": metrics["Precision"],
        "Recall": metrics["Recall"],
        "F1-score": metrics["F1-score"]
    })

individual_models_df = pd.DataFrame(model_results_list)

individual_models_df = individual_models_df.sort_values(by='F1-score', ascending=False).reset_index(drop=True)

print("-------------------Models Performance-------------------")
display(individual_models_df)

-------------------Models Performance-------------------


,Model,Accuracy,Precision,Recall,F1-score
0,Hard Voting,0.973684,0.972222,0.985915,0.979021
1,Logistic Regression,0.973684,0.972222,0.985915,0.979021
2,Soft Voting,0.973684,0.972222,0.985915,0.979021
3,Naive Bayes,0.964912,0.958904,0.985915,0.972222
4,AdaBoost,0.964912,0.958904,0.985915,0.972222
5,Random Forest,0.964912,0.958904,0.985915,0.972222
6,SGD Classifier,0.964912,0.985507,0.957746,0.971429
7,Bagging,0.956140,0.958333,0.971831,0.965035
8,Decision Tree,0.947368,0.957746,0.957746,0.957746
9,KNN,0.947368,0.957746,0.957746,0.957746


# Part 4 – Voting Classifier

In [29]:
# TODO: Import VotingClassifier
from sklearn.ensemble import VotingClassifier

# TODO: Create hard voting classifier using at least 3 base models
voting_hard = VotingClassifier(
    estimators=[
        ('dt', dt_model),
        ('nb', nb_model),
        ('knn', knn_model),
        ('lr', lr_model),
        ('sgd', sgd_model)
    ],
    voting='hard'
)

# TODO: Fit model
voting_hard.fit(X_train_scaled, y_train)

# TODO: Predict
vh_preds = voting_hard.predict(X_test_scaled)

# TODO: Compute evaluation metrics
vh_results = {
    "Accuracy": accuracy_score(y_test, vh_preds),
    "Precision": precision_score(y_test, vh_preds),
    "Recall": recall_score(y_test, vh_preds),
    "F1-score": f1_score(y_test, vh_preds),
    "Confusion Matrix": confusion_matrix(y_test, vh_preds)
}

# TODO: Store results
results_dict["Hard Voting"] = vh_results

In [30]:
# Re-initializing SGD for Soft Voting compatibility
sgd_soft = SGDClassifier(loss='log_loss', random_state=42)

# TODO: Create soft voting classifier
voting_soft = VotingClassifier(
    estimators=[
        ('dt', dt_model),
        ('nb', nb_model),
        ('knn', knn_model),
        ('lr', lr_model),
        ('sgd', sgd_soft) # version supports probabilities
    ],
    voting='soft'
)

# NOTE: All models used must support predict_proba
# TODO: Fit model
voting_soft.fit(X_train_scaled, y_train)

# TODO: Predict
vs_preds = voting_soft.predict(X_test_scaled)

# TODO: Compute evaluation metrics
vs_results = {
    "Accuracy": accuracy_score(y_test, vs_preds),
    "Precision": precision_score(y_test, vs_preds),
    "Recall": recall_score(y_test, vs_preds),
    "F1-score": f1_score(y_test, vs_preds),
    "Confusion Matrix": confusion_matrix(y_test, vs_preds)
}

# TODO: Store results
results_dict["Soft Voting"] = vs_results

In [31]:
print(vh_results)

{'Accuracy': 0.9736842105263158, 'Precision': 0.9722222222222222, 'Recall': 0.9859154929577465, 'F1-score': 0.9790209790209791, 'Confusion Matrix': array([[41,  2],
       [ 1, 70]])}


In [32]:
print(vs_results)

{'Accuracy': 0.9736842105263158, 'Precision': 0.9722222222222222, 'Recall': 0.9859154929577465, 'F1-score': 0.9790209790209791, 'Confusion Matrix': array([[41,  2],
       [ 1, 70]])}


In [33]:
print("--- Hard Voting Classifier Results ---")
for metric, value in results_dict["Hard Voting"].items():
    if metric != "Confusion Matrix":
        print(f"{metric}: {value:.4f}")
    else:
        print(f"{metric}:\n{value}")

print("\n" + "="*40 + "\n")

print("--- Soft Voting Classifier Results ---")
for metric, value in results_dict["Soft Voting"].items():
    if metric != "Confusion Matrix":
        print(f"{metric}: {value:.4f}")
    else:
        print(f"{metric}:\n{value}")

--- Hard Voting Classifier Results ---
Accuracy: 0.9737
Precision: 0.9722
Recall: 0.9859
F1-score: 0.9790
Confusion Matrix:
[[41  2]
 [ 1 70]]


--- Soft Voting Classifier Results ---
Accuracy: 0.9737
Precision: 0.9722
Recall: 0.9859
F1-score: 0.9790
Confusion Matrix:
[[41  2]
 [ 1 70]]


### In 2–3 sentences:
What is the difference between hard voting and soft voting?

- Hard voting predicts the final class by taking the average highest-class count, meaning it simply follows the majority choice of the individual classifiers

- Soft voting, on the other hand, calculates the average of the class probabilities and gives more weight to highly confident votes; as seen in the example, this can lead to a different result if one classifier is very certain about a specific outcome.

# Part 5 – Bagging and Boosting

In [34]:
# TODO: Import BaggingClassifier
from sklearn.ensemble import BaggingClassifier

# TODO: Initialize Bagging model (choose base estimator)
bag_model = BaggingClassifier(random_state=42)

# TODO: Fit model
bag_model.fit(X_train_scaled, y_train)

# TODO: Predict
bag_preds = bag_model.predict(X_test_scaled)

# TODO: Compute evaluation metrics
bag_results = {
    "Accuracy": accuracy_score(y_test, bag_preds),
    "Precision": precision_score(y_test, bag_preds),
    "Recall": recall_score(y_test, bag_preds),
    "F1-score": f1_score(y_test, bag_preds),
    "Confusion Matrix": confusion_matrix(y_test, bag_preds)
}

# TODO: Store results
results_dict["Bagging"] = bag_results

In [36]:
# TODO: Import RandomForestClassifier
from sklearn.ensemble import RandomForestClassifier

# TODO: Initialize model
rf_model = RandomForestClassifier(random_state=42)

# TODO: Fit model
rf_model.fit(X_train_scaled, y_train)

# TODO: Predict
rf_preds = rf_model.predict(X_test_scaled)

# TODO: Compute evaluation metrics
rf_results = {
    "Accuracy": accuracy_score(y_test, rf_preds),
    "Precision": precision_score(y_test, rf_preds),
    "Recall": recall_score(y_test, rf_preds),
    "F1-score": f1_score(y_test, rf_preds),
    "Confusion Matrix": confusion_matrix(y_test, rf_preds)
}

# TODO: Store results
results_dict["Random Forest"] = rf_results

In [37]:
# TODO: Import AdaBoostClassifier
from sklearn.ensemble import AdaBoostClassifier

# TODO: Initialize model
ada_model = AdaBoostClassifier(random_state=42)

# TODO: Fit model
ada_model.fit(X_train_scaled, y_train)

# TODO: Predict
ada_preds = ada_model.predict(X_test_scaled)

# TODO: Compute evaluation metrics
ada_results = {
    "Accuracy": accuracy_score(y_test, ada_preds),
    "Precision": precision_score(y_test, ada_preds),
    "Recall": recall_score(y_test, ada_preds),
    "F1-score": f1_score(y_test, ada_preds),
    "Confusion Matrix": confusion_matrix(y_test, ada_preds)
}

# TODO: Store results
results_dict["AdaBoost"] = ada_results

# Part 6 – Model Comparison

In [40]:
# TODO: Create a comparison DataFrame that includes:

# Model Name
# Accuracy
# Precision
# Recall
# F1-score
comparison_list = []
for model_name, metrics in results_dict.items():
    row = metrics.copy()
    row['Model Name'] = model_name
    row.pop('Confusion Matrix', None)
    comparison_list.append(row)

comparison_df = pd.DataFrame(comparison_list)

comparison_df = comparison_df[['Model Name', 'Accuracy', 'Precision', 'Recall', 'F1-score']]

# TODO: Sort results by F1-score (descending)
comparison_df = comparison_df.sort_values(by='F1-score', ascending=False)

# TODO: Display the table
print(comparison_df.to_string(index=False))

         Model Name  Accuracy  Precision   Recall  F1-score
        Hard Voting  0.973684   0.972222 0.985915  0.979021
Logistic Regression  0.973684   0.972222 0.985915  0.979021
        Soft Voting  0.973684   0.972222 0.985915  0.979021
        Naive Bayes  0.964912   0.958904 0.985915  0.972222
           AdaBoost  0.964912   0.958904 0.985915  0.972222
      Random Forest  0.964912   0.958904 0.985915  0.972222
     SGD Classifier  0.964912   0.985507 0.957746  0.971429
            Bagging  0.956140   0.958333 0.971831  0.965035
      Decision Tree  0.947368   0.957746 0.957746  0.957746
                KNN  0.947368   0.957746 0.957746  0.957746


# Part 7 – Metric Selection and Interpretation

Answer the following:

1. Which model achieved the highest accuracy?
   - Hard Voting, Logistic Regression and Soft Voting all tied for the highest accuracy at approximately 0.9737.

2. Which model achieved the highest recall?
   - Multiple models tied for the highest recall of 0.9859, including Hard Voting, Logistic Regression, Soft Voting, Naive Bayes, AdaBoost, and Random Forest.

3. Which model achieved the highest F1-score?
   - Hard Voting, Logistic Regression, and Soft Voting tied for the highest F1-score at 0.9790.

4. In the context of breast cancer diagnosis:
   - Which metric is most important?
     - Recall is the most important metric in this case.
   - Why?
     - High recall means the model is successful at identifying the majority of actual positive (malignant) cases. Especially, in medical diagnostics, a False Negative (failing to detect cancer when it is present) is much more dangerous than a False Positive, as it results in a patient not receiving life-saving treatment (I learned this by heart since Dr. Ezzeddine talked about this so many times in class).
5. If a model has high accuracy but low recall, what does that imply in this medical context?
   - It implies the model is very good at identifying healthy (benign) patients but frequently misses patients who actually have cancer. While the model looks "good" on paper because the overall percentage of correct guesses is high, it is medically unreliable and dangerous because it provides a false sense of security to sick individuals.

# Part 8 – Final Reflection (Required)

## Final Reflection (1-2 paragraphs)

Write a short summary in your own words explaining:

- What you observed when comparing individual models vs ensemble models.
- Whether ensemble methods improved performance.
- Which model you would choose for real-world deployment.
- Which metric best represents performance in this case study and why.
- Any surprising findings.


### Your Reflection:

TODO: Write your summary here.

- In this lab, I observed that individual models like Logistic Regression and Naive Bayes provided highly competitive results, but Ensemble Methods generally offered more robust and stable performance. Specifically, Hard and Soft Voting classifiers tied for the top accuracy (0.9737) and F1-score (0.9790) by aggregating the predictions of all five base learners. This demonstrates that ensembles can mitigate the weaknesses of single algorithms, such as the lower recall seen in the Decision Tree or KNN, by balancing them with the higher sensitivity of models like Logistic Regression.

- For real-world deployment, I would choose the Soft Voting Classifier. While it tied with Logistic Regression on several metrics, soft voting is theoretically superior because it weighs the "confidence" (probabilities) of each model rather than just the final class count. In this case study, Recall is the most critical metric because a "False Negative"—missing a malignant diagnosis—carries a much higher human cost than a False Positive. Accuracy alone can be misleading here, a model could have high accuracy by simply predicting the majority class correctly while still missing the critical, less frequent malignant cases. It was surprising that Logistic Regression matched the performance of complex ensembles, but the ensemble remains the safer choice for deployment as it is typically more resilient to data variations.